src/distributions/range.rs

```diff
@@ -47,126 +45,451 @@ use distributions::Distribution;
+    scale: X,
+    offset: X,
+}
+
+macro_rules! range_float_impl {
+    ($ty:ty, $bits_to_discard:expr, $next_u:ident) => {
         impl SampleRange for $ty {
-            fn construct_range(low: $ty, high: $ty) -> Range<$ty> {
-                Range {
-                    low: low,
-                    range: high - low,
-                    accept_zone: 0.0 // unused
+            type T = RangeFloat<$ty>;
+        }
+
+        impl RangeImpl for RangeFloat<$ty> {
+            type X = $ty;
+
+            fn new(low: Self::X, high: Self::X) -> Self {
+                let scale = high - low;
+                let offset = low - scale;
+                RangeFloat {
+                    scale: scale,
+                    offset: offset,
                 }
             }
-            fn sample_range<R: Rng + ?Sized>(r: &Range<$ty>, rng: &mut R) -> $ty {
-                r.low + r.range * rng.gen::<$ty>()
+
+            fn new_inclusive(low: Self::X, high: Self::X) -> Self {
+                // Same as `new`, because the boundaries of a floats range are
+                // (at least currently) not exact due to rounding errors.
+                RangeImpl::new(low, high)
+            }
+
+            fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> Self::X {
+                // Generate a value in the range [1, 2)
+                let value1_2 = (rng.$next_u() >> $bits_to_discard)
+                               .into_float_with_exponent(0);
+                // Doing multiply before addition allows some architectures to
+                // use a single instruction.
+                value1_2 * self.scale + self.offset
             }
         }
     }
 }
 
-float_impl! { f32 }
-float_impl! { f64 }
+range_float_impl! { f32, 32 - 23, next_u32 }
+range_float_impl! { f64, 64 - 52, next_u64 }
+
 
 #[cfg(test)]
 mod tests {
-    use distributions::Distribution;
-    use super::Range as Range;
+    use Rng;
+    use distributions::{Distribution};
+    use distributions::range::{Range, RangeImpl, RangeFloat, SampleRange};
+
+    #[test]
+    fn test_fn_range() {
+        let mut r = ::test::rng(814);
+        for _ in 0..1000 {
+            let a = Range::new(-3, 42).sample(&mut r);
+            assert!(a >= -3 && a < 42);
+            assert_eq!(Range::new(0, 1).sample(&mut r), 0);
+            assert_eq!(Range::new(-12, -11).sample(&mut r), -12);
+        }
+
+        for _ in 0..1000 {
+            let a = Range::new(10, 42).sample(&mut r);
+            assert!(a >= 10 && a < 42);
+            assert_eq!(Range::new(0, 1).sample(&mut r), 0);
+            assert_eq!(Range::new(3_000_000, 3_000_001).sample(&mut r), 3_000_000);
+        }
+    }
+
+    #[test]
+    #[should_panic]
+    fn test_fn_range_panic_int() {
+        let mut r = ::test::rng(815);
+        Range::new(5, -2).sample(&mut r);
+    }
+
+    #[test]
+    #[should_panic]
+    fn test_fn_range_panic_usize() {
+        let mut r = ::test::rng(816);
+        Range::new(5, 2).sample(&mut r);
```

Comment:
Doesn't use `usize` like name implies. This and the previous fn are redundant. Maybe add one unit-test to test all supported int types?